pkg/stream/xprotocol/keepalive.go

```diff
@@ -102,10 +108,70 @@ func NewKeepAlive(codec str.Client, proto api.XProtocol, timeout time.Duration)
 	return kp
 }
 
+// NewKeepAliveWithConfig creates a keepalive object with keepalive config
+func NewKeepAliveWithConfig(codec str.Client, proto api.XProtocol, conf types.KeepAliveConfig) types.KeepAlive {
+	if conf.Timeout <= 0 {
+		conf.Timeout = time.Second // set default timeout
+	}
+	kp := &xprotocolKeepAlive{
+		Codec:     codec,
+		Protocol:  proto,
+		Timeout:   conf.Timeout,
+		Interval:  conf.Interval,
+		afterFunc: time.AfterFunc,
+		Callbacks: make([]types.KeepAliveCallback, 0),
+		stop:      make(chan struct{}),
+		requests:  make(map[uint64]*keepAliveTimeout),
+	}
+
+	// initially set previous heartbeat request success
+	kp.previousIsSucc.Store(true)
+
+	// initially setup sets the heartbeat check fast failure task to nil
+	kp.fastFailTask.Store(nilTask)
+
+	// register keepalive to connection event listener
+	// if connection is closed, keepalive should stop
+	kp.Codec.AddConnectionEventListener(kp)
+	return kp
+}
+
+func (kp *xprotocolKeepAlive) StartSchedule() {
+	if kp.Interval <= 0 {
+		return
+	}
+	if log.DefaultLogger.GetLogLevel() >= log.DEBUG {
+		log.DefaultLogger.Debugf("[stream] [xprotocol] [keepalive] [start schedule] conn id:[%d], interval: %v",
+			kp.Codec.ConnID(), kp.Interval)
+	}
+	kp.runSchedule()
+}
+
+func (kp *xprotocolKeepAlive) runSchedule() {
+	kp.afterFunc(kp.Interval, func() {
+		defer func() {
+			if r := recover(); r != nil {
+				log.DefaultLogger.Alertf("upstream.keepalive", "[stream] [xprotocol] [keepalive] "+
+					"Schedule send keepalive heartbeat conn id:[%d], panic %v\n%s",
+					kp.Codec.ConnID(), r, string(debug.Stack()))
```

Comment:
```suggestion
					kp.Codec.ConnID(), r, debug.Stack())
```
We don't need to call `string` when we already used `%s` in the format